src/ai/rival-team-gen.ts

```diff
@@ -0,0 +1,331 @@
+      ) {
+        choices.push(i);
+      } else if (typeof entry !== "number") {
+        for (const [j, subEntry] of entry.entries()) {
+          if (checkTypingConstraints(subEntry, existingTypes, existingWeaknesses, balanceTypes, balanceWeaknesses)) {
+            choices.push([i, j]);
+          }
+        }
+      }
+    }
+  }
+
+  if (choices.length === 0) {
+    for (const [i, entry] of pool.entries()) {
+      if (typeof entry === "number") {
+        choices.push(i);
+      } else {
+        for (const j of entry.keys()) {
+          choices.push([i, j]);
+        }
+      }
+    }
+  }
+
+  return choices;
+}
+
+/**
+ * Randomly selects one of the `Species` from `speciesPool`, determines its evolution, level, and strength.
+ * Then adds Pokemon to `globalScene`.
+ * @param config - The configuration for the rival pool fight
+ * @param slot - The slot being generated for (0-5)
+ * @param referenceConfig - (Default {@linkcode RIVAL_6_POOL}); The final rival pool configuration to use if `config` is `RIVAL_POOL_CONFIG.FINAL`
+ *
+ * @throws
+ * If no configuration is found for the specified slot.
+ */
+export function getRandomRivalPartyMemberFunc(
+  config: RivalPoolConfig,
+  slot: number,
+  referenceConfig: RivalPoolConfig = RIVAL_6_POOL,
+): (level: number, strength: PartyMemberStrength) => EnemyPokemon {
+  // Protect against out of range slots.
+  // Only care about this in dev to be caught during development; it will be excluded in production builds.
+  if (import.meta.env.DEV && slot > config.length) {
+    throw new Error(`Slot ${slot} is out of range for the provided config of length ${config.length}`);
+  }
+  return (level: number, _strength: PartyMemberStrength) => {
+    const { pool, postProcess, balanceTypes, balanceWeaknesses } = config[slot];
+
+    const existingTypes = new Map<PokemonType, number>();
+    const existingWeaknesses = new Map<PokemonType, number>();
+
+    if (slot === 0) {
+      // Clear out the rolls from previous rival generations
+      CHOSEN_RIVAL_ROLLS.fill(undefined);
+    } else if (balanceTypes || balanceWeaknesses) {
+      calcPartyTypings(slot, existingTypes, existingWeaknesses, balanceTypes, balanceWeaknesses, referenceConfig);
+    }
+
+    // Filter the pool to its choices, or map it
+
+    let species: SpeciesId | SpeciesId[];
+
+    // When converting pool to choices, base off of the reference config
+    // to use for type balancing, as we only narrow based on what the slot
+    // will be in its final stage
+    const choices = convertPoolToChoices(
+      referenceConfig[slot].pool,
+      existingTypes,
+      existingWeaknesses,
+      balanceTypes,
+      balanceWeaknesses,
+    );
+
+    const choice = randSeedItem(choices);
+    if (typeof choice === "number") {
+      species = pool[choice] as SpeciesId;
+      CHOSEN_RIVAL_ROLLS[slot] = [choice];
+    } else {
+      species = (pool[choice[0]] as SpeciesId[])[choice[1]];
```

Comment:
This typecast doesn't seem necessary (anymore). Didn't get any type errors when removing it.
```suggestion
      species = pool[choice[0]][choice[1]];
```